docs/design.md

```diff
@@ -0,0 +1,279 @@
+
+```go
+// Example of internal data structure of an index of data provider.
+type AdvertisementID Index_Link
+
+type Index struct {
+        Previous: Index_Link{},
+        Entries: []Entry{
+                IsRm:           bool, 
+                Cids:           []cid.Cid,        // NOTE: This can also be string.
+                Metadata:       []byte,
+        },
+} 
+```
+
+_NOTE: We need to build the selector or the IPLD storer in the indexer node so that it checks the latest Advertisement
+seen and it terminates Graphsync's DAG traversal conveniently._
+
+### IPLD-unaware Ingestion 
+> Out of scope of MVP. For the MVP we are only considering IPLD-aware ingestion, and we enforce the use of IPLD to providers.
+
+It implements a libp2p request-response ingestion protocol that indexer nodes can use when interacting with providers that are not indexing their
+data using IPLD, i.e. Advertisements are not directly linked to data. This protocol is inspired by 
+[Filecoin's Block Sync protocol](https://spec.filecoin.io/#section-algorithms.block_sync). The underlaying idea is the same, to "sync a chain of data".
+
+Thus, when calling `Sync(ctx, p, latest)`, indexer node sends a first request for ingestion as:
+```go
+req = IngestionRequest{
+        ID: latest,
+        Start: 0, 
+}
+```
+
+Provider will answer with the list of entries for `latest` pointing to the previous `AdvertisementID` in the `Advertisement` field. `Size` specifies
+the total number of entries that the advertisement include to give the indexer knowledge of when all entries for the update have been received.
+
+```go
+resp = IngestionResponse {
+        Size: 2
+        Error: 0
+        Start: 0
+        Advertisement: Advertisement{
+                ID: latest,
+                Previous: "latest-1",
+                IndexerID: Link_Index,
+                Provider: p,
+                Signature: []byte{},  // The original signature of the advertisement.
+                GraphSupport: false, 
+        }
+        Entries: {
+                []Entry{
+                        Cids: {Cid1, Cid2, ...},
+                        Metadata: []byte{},
+                }
+}
+```
+- If `Size > len(Entries)`, the indexer-node sends a new request to keep fetching data for `latest` with `Start: 1`
+until all entries for the Advertisement have been received. It then inspects `Advertisement` to see if `Previous` is equal
+to the latest `AdvertisementID` seen for that provider, if this is not the case, it sends a new request with `latest-1`
+and repeats the process over and over again until the `AdvertisementID` seen in `Previous` for the latest response
+equals the latest one seen by indexer node for that data provider and the full sync for the missing range is complete.
+
+Note that pagination is performed at an `Entry` level, i.e. there is no intra-Entry pagination (at least not initially). The data provider should be responsible for dividing single Entries with a large number of CIDs into
```

Comment:
The design doc is very specific about how [response pagination](https://www.notion.so/protocollabs/Indexer-Node-Design-4fb94471b6be4352b6849dc9b9527825#278589e9c95d4363a5986b4309fc0f5a) works.  Maybe consider using  that text, or something equivalent.